File: simple-stack-example-mvvm-fragments/src/main/java/com/zhuinden/simplestackexamplemvvm/application/BaseFragment.java
package com.zhuinden.simplestackexamplemvvm.application;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;

/**
 * Created by Zhuinden on 2017.07.26..
 */

public abstract class BaseFragment<VM>
        extends Fragment {
    @NonNull
    public final <T extends BaseKey> T getKey() {
        Bundle args = getArguments();
        if(args == null) {
            throw new IllegalStateException("Fragment cannot have null arguments.");
        }
        T key = args.getParcelable("KEY");
        if(key == null) {
            throw new IllegalStateException("Fragment cannot have null key");
        }
        return key;
    }

    public abstract void bindViewModel(VM viewModel);
}

File: simple-stack-example-kotlin/src/main/kotlin/com/zhuinden/simplestackexamplekotlin/BaseFragment.kt
package com.zhuinden.simplestackexamplekotlin

import android.support.v4.app.Fragment

/**
 * Created by Owner on 2017. 06. 29..
 */

open class BaseFragment : Fragment() {
    fun <T : BaseKey> getKey(): T? = arguments?.getParcelable<T>("KEY")
}
